server/util/networking/networking.go

```diff
@@ -642,31 +590,92 @@ func (s cleanupStack) Cleanup(ctx context.Context) error {
 	return nil
 }
 
-// ConfigureNATForTapInNamespace configures NAT in the namespace so that
-// outgoing IP packets from the tap device in the namespace are translated to
-// the namespaced veth device IP, and incoming IP packets to the tap device are
-// translated to the VM tap device IP.
+// VMNetwork represents a fully-provisioned VM network, which consists of a net
+// namespace, virtual network interfaces, and associated host configuration.
 //
-// Since the host-side of the veth pair also has NAT configured, this allows the
-// VM to communicate with external networks.
-func ConfigureNATForTapInNamespace(ctx context.Context, vethPair *VethPair, vmIP string) error {
-	if err := runCommand(ctx, namespace(vethPair.netns, "iptables", "--wait", "-t", "nat", "-A", "POSTROUTING", "-o", vethPair.namespacedDevice, "-s", vmIP, "-j", "SNAT", "--to", vethPair.network.NamespacedIP())...); err != nil {
-		return err
+// Deleting a VM network deletes the net namespace as well as all associated
+// resources, and reverts the applied host configuration.
+type VMNetwork struct {
+	netns    *Namespace
+	vethPair *vethPair
+	cleanup  func(ctx context.Context) error
+}
+
+// CreateVMNetwork initializes a network namespace, networking
+// interfaces, and host configuration required for VM networking.
+func CreateVMNetwork(ctx context.Context, tapDeviceName, tapAddr, vmIP string) (_ *VMNetwork, err error) {
+	var cleanupStack cleanupStack
+	defer func() {
+		// If we failed to fully set up the network, make sure to clean up any
+		// resources that were partially set up.
+		if err != nil {
+			_ = cleanupStack.Cleanup(ctx)
+		}
+	}()
+
+	// Create a net namespace.
+	netns, err := createUniqueNetNamespace(ctx)
+	if err != nil {
+		return nil, status.WrapError(err, "create net namespace")
 	}
-	if err := runCommand(ctx, namespace(vethPair.netns, "iptables", "--wait", "-t", "nat", "-A", "PREROUTING", "-i", vethPair.namespacedDevice, "-d", vethPair.network.NamespacedIP(), "-j", "DNAT", "--to", vmIP)...); err != nil {
-		return err
+	cleanupStack = append(cleanupStack, func(ctx context.Context) error {
+		return netns.Delete(ctx)
+	})
+
+	// Create a veth pair with one end in the namespace.
+	vethPair, err := setupVethPair(ctx, netns)
```

Comment:
I'm not very familiar with the setup - just basing off the old code. Does it matter that we're now doing this before creating the tap device?

Reply:
Nope, the order of creating the interfaces doesn't matter - the tap / veth devices are independent except for the NAT rule that translates the tap IP to/from the namespaced veth IP, and that NAT rule setup will succeed as long as the namespaced veth device exists when we run it